_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,392 @@
+
+1. Unsafe Integer Methods: Prove the safety of methods like `unchecked_add`, `unchecked_sub`, `unchecked_mul`, `unchecked_shl`, `unchecked_shr`, and `unchecked_neg` for various integer types.
+2. Safe API Verification: Verify safe APIs that leverage the unsafe integer methods from Part 1, such as `wrapping_shl`, `wrapping_shr`, `widening_mul`, and `carrying_mul`.
+3. Float to Integer Conversion: Verify the absence of undefined behavior in `to_int_unchecked` for all floating-point types: `f16`, `f32`, `f64`, and `f128`.
+
+In addition to verifying the methods under their specified safety preconditions, we needed to ensure the absence of specific [undefined behaviors](https://github.com/rust-lang/reference/blob/142b2ed77d33f37a9973772bd95e6144ed9dce43/src/behavior-considered-undefined.md):
+
+- Invoking undefined behavior via compiler intrinsics.
+- Reading from uninitialized memory.
+- Mutating immutable bytes.
+- Producing an invalid value.
+
+## Approach
+
+To tackle this challenge, we utilized [Kani](https://github.com/model-checking/kani)'s verification capabilities to write proof harnesses for unsafe methods. Our strategy involved:
+
+1. **Specifying Safety Preconditions**: To ensure the methods behaved correctly, we explicitly specified safety preconditions in the code. These preconditions were used to:
+
+   - Guide Kani's input generation.
+   - Define the boundaries for valid inputs.
+   - Avoid triggering undefined behavior in cases where the preconditions were violated.
+2. **Writing Verification Harnesses**: We developed Kani proof harnesses for each unsafe method by:
+
+   - Generating inputs that satisfy the required safety preconditions.
+   - Invoking the methods within an `unsafe` block to verify their correctness and safety under the defined conditions.
+3. **Using Macros for Code Generation**: To efficiently handle the large number of methods and types that required verification, we utilized Rust macros. These macros automated the generation of proof harnesses, reducing redundancy and improving maintainability.
+4. **Handling Large Input Spaces**: For methods with large input spaces (e.g., `unchecked_mul` for 64-bit integers), we made the verification process more tractable by:
+
+   - Introducing assumptions to limit the range of inputs.
+   - Focusing on specific edge cases (e.g., maximum values, zero, and boundary conditions) that are most likely to cause overflow or underflow.
+
+We will walk through each part with examples from the Unsafe Integer Methods part of the challenge.
+
+## Implementation
+
+### 1. Specifying Safety Preconditions
+
+The following code demonstrates how to specify safety preconditions for an unsafe numeric method `unchecked_add` using [**function contracts**](https://github.com/model-checking/kani/blob/main/rfc/src/rfcs/0009-function-contracts.md) to ensure correct behavior under the stated conditions.
```

Comment:
```suggestion
The following code demonstrates how to specify safety preconditions for an unsafe numeric method `unchecked_add` using [**function contracts**](https://model-checking.github.io/kani/reference/experimental/contracts.html) to ensure correct behavior under the stated conditions.
```
Let's use our documentation from our book instead of the RFC.